assert: multiple improvements

BridgeAR opened the pull request:
<s>I am going to split this into smaller commits. Therefore it is WIP.</s>

The following changes are in here:

- Switched `+ / -` | `red / green` (by request). It does seem to feel a bit more natural this way. Let's hope that most people agree with this.
- Short primitives do not use the diff anymore (by request). Especially short numbers can be read well like `1 !== 2`.
- Improved error descriptions. It was not really clear how things worked. It should now be clear what everything is about.
- Added a position indicator for single lines that fit into the tty.
- <s>Simple assert will now rarely bail out when checking the call site and the output got improved.</s>
- <s>Some performance optimizations</s>
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

jasnell commented on lib/assert.js:
```diff
@@ -128,8 +128,10 @@ function getBuffer(fd, assertLine) {
   const bytesPerRead = 8192;
   const buffers = [];
+  // Use a single buffer up front that is reused until the call site is found.
+  let buffer = Buffer.allocUnsafe(bytesPerRead);
```
💯

jasnell commented on lib/assert.js:
```diff
@@ -152,10 +154,65 @@ function getBuffer(fd, assertLine) {
+      nodes = parseExpressionAt(code, start);
+      if (nodes.type !== 'CallExpression') {
+        // Walk the tree to find the call expression.
+        const walk = require('internal/deps/acorn/dist/walk');
```
Perhaps make this a more "traditional" lazy load by having a top level `let walk;` then here `walk = walk || require('internal/deps/acorn/dist/walk')` or something similar?

devsnek commented:
it might help to just remove the +/- and use colors. i generally try to just focus on the colors because the +/- add another dimension that confuses the heck outta me.

BridgeAR commented:
@devsnek using only colors is not an option in case people do not have a TTY with color support. In the end the color is just the same as + / - and only a different notation for the same thing.

BridgeAR commented:
I moved the simple assert changes out of this PR. I guess the rest is fine to be in one commit.

So PTAL.

Trott commented:
@nodejs/tsc @nodejs/testing

targos commented on lib/internal/assert.js:
```diff
@@ -10,13 +10,18 @@ let green = '';
-  strictEqual: 'Input A expected to strictly equal input B',
-  notStrictEqual: 'Input A expected to strictly not equal input B'
+const kReadableOperator = {
+  deepStrictEqual: 'Expected input to be strictly deep-equal',
```
suggestion: "inputs" instead of "input"

BridgeAR replied:
I am good either way as it seems that both would work. @Trott @vsemozhetbyt do you have a preference?

vsemozhetbyt replied:
For me, "inputs" seems more clear (no confusing "input is deep-equal — to what?"), but I am not a native speaker)

Trott replied:
I agree with @targos and @vsemozhetbyt for the reason offered by @vsemozhetbyt (using "inputs" removes potential ambiguity).

targos commented on lib/internal/assert.js:
```diff
@@ -10,13 +10,18 @@ let green = '';
-  notStrictEqual: 'Input A expected to strictly not equal input B'
+const kReadableOperator = {
+  deepStrictEqual: 'Expected input to be strictly deep-equal',
+  notDeepStrictEqual: 'Expected "actual" not to be strictly deep-equal',
```
why use "actual" here?

BridgeAR replied:
It is the value of the input argument "actual". I had hoped that this is actually clearer than the version before. Do you have a idea or a suggestion for to make it even better?

targos replied:
I'm wondering why it's not `input` or `inputs` like with the sibling method

targos replied:
I see. Thanks for the explanation, this seems fine.

BridgeAR commented:
Comment addressed.

CI https://ci.nodejs.org/job/node-test-pull-request/15877/

mcollina commented on lib/internal/assert.js:
```diff
@@ -49,22 +54,46 @@ function inspectValue(val) {
+          (actual !== 0 || expected !== 0)) { // -0 === +0
+        return `${kReadableOperator[operator]}:\n\n` +
+            `${actualLines[0]} !== ${expectedLines[0]}\n`;
+      }
```
Can you please add a comment explaining this block of logic?

BridgeAR replied:
Done.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/15896/

addaleax commented on lib/internal/assert.js:
```diff
@@ -49,22 +54,53 @@ function inspectValue(val) {
+    // If the stdout is a tty and the input length is lower than the current
+    // columns per line, add a mismatch indicator below the output.
+    } else if (process.stdout.isTTY &&
+               inputLength < process.stdout.columns) {
```
I’d prefer not to add a dependency on `process.stdout` here … there’s no guarantee that this is where the error output ends up, and it makes `assert` depend on something it shouldn’t depend on (or makes it more so, at least).

Can we pick a value, e.g. 80 or 100 or so?

BridgeAR replied:
I am aware that there is no guarantee that this is where the error output ends up. It is also the reason why I only added the indicator if it is a tty (and yes, even then the error could end up somewhere else but that is unlikely).

If I understand you correct you would actually like to have the indicator all the time? In that case I could add a default of 80 in case it is not a tty and check stdout otherwise?

That all aside: I would actually like to add a proper diffing algorithm at some point with which it would be possible to use colors to highlight the differences instead. But that is a bit more work and I got quite some pushback when it comes to improvements to assert, so I am not really sure if that is something worth trying.

BridgeAR commented:
@addaleax I added a default of 80 for all cases where stdout is not a tty.

@targos I also added a "to" to prevent the make things even clearer (it addresses https://github.com/nodejs/node/pull/21628#discussion_r202513165).

I also moved some code up to remove a intermediate variable and to shorten the code path in that case.

CI https://ci.nodejs.org/job/node-test-pull-request/16029/

BridgeAR commented:
I just changed two more small things:

1) Instead of checking stdout, it will now check stderr as it is more likely that the error ends up there.
2) The indicator is ignored for the first three characters.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/16038/

BridgeAR commented:
I just pushed the doc changes that were missing. The CI was green after resuming as it can be seen by the green check mark next to the commit before the docs. Since the doc changes do not need a full CI, I did not start anything else anymore.

PTAL. I think this is ready otherwise and could land.

BridgeAR commented:
PTAL. It would be nice to get LG for the last changes.

BridgeAR commented:
@nodejs/documentation PTAL

vsemozhetbyt commented:
Just a nit question. Sometimes in the error logs we have `AssertionError [ERR_ASSERTION]: ...` and sometimes just `AssertionError: ...`. Is there some system in this?

BridgeAR commented:
The line length was to long when adding the `ERR_ASSERTION` every time. Therefore it was some times removed.

BridgeAR commented:
Thanks a lot.

Landed in 0518b9edf33bfffac53ac5f706694a205ff754a2 :tada:
